src/devices/machine/tk10_keyboard.cpp

```diff
@@ -0,0 +1,256 @@
+// license:BSD-3-Clause
+// copyright-holders:Golden Child
+/**********************************************************************
+
+    TK-10 Apple ][ Clone keyboard
+
+
+
+to patch into the apple2 driver:
+
+
++#include "machine/tk10_keyboard.h"
+
++void tk10_data_write(u8 data);
+
++void apple2_state::tk10_data_write(u8 data)
++{
++ //    printf("tk10_code_sent_to_apple %x\n",data);
++       if (data & 0x80)
++       {
++                m_transchar = data & 0x7f;
++                m_strobe = 0x80;
++       }
++}
++
+
++       tk10_keyboard_device &tk10(TK10_KEYBOARD(config, "tk10", 0));
++       tk10.data_write_cb().set(FUNC(apple2_state::tk10_data_write));
++
+
+
+
+**********************************************************************/
+
+#include "emu.h"
+
+#include "tk10_keyboard.h"
+
+#include "utf8.h"
+
+
+//**************************************************************************
+//  DEVICE DEFINITION
+//**************************************************************************
+
+// device type definition
+DEFINE_DEVICE_TYPE(TK10_KEYBOARD, tk10_keyboard_device, "tk10_keyboard", "TK-10 Keyboard")
+
+
+//-------------------------------------------------
+//  tk10_keyboard_device - constructor
+//-------------------------------------------------
+
+tk10_keyboard_device::tk10_keyboard_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, TK10_KEYBOARD, tag, owner, clock)
+	, m_data_write(*this)
+	, m_reset_write(*this)
+	, m_tk10cpu(*this, "tk10cpu")
+	, m_keyboard(*this, "P2%u", 0U)
+	, m_kbspecial(*this, "kbspecial")
+{
+}
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+
+void tk10_keyboard_device::device_start()
+{
+}
+
+//-------------------------------------------------
+//  device_add_mconfig - device-specific config
+//-------------------------------------------------
+
+void tk10_keyboard_device::device_add_mconfig(machine_config &config)
+{
+	I8039(config, m_tk10cpu, 6_MHz_XTAL);
+	m_tk10cpu->set_addrmap(AS_PROGRAM, &tk10_keyboard_device::tk10_mem_map);
+
+	m_tk10cpu->t1_in_cb().set([this] () -> u8
+	{
+		u8 kbspecial = m_kbspecial->read();  // left and right shift keys
+		return (BIT(kbspecial, 0) & BIT(kbspecial, 1));
+	});
+
+	m_tk10cpu->t0_in_cb().set([this] () -> u8
+	{
+		return (BIT(m_kbspecial->read(), 2));  // control key
+	});
+
+	m_tk10cpu->p1_in_cb().set([this] () -> u8
+	{
+		for (int i = 0; i < 8; i++)
+		{
+			if (!BIT(m_tk10cpu->p2_r(), i)) return m_keyboard[i]->read();
+			// read key matrix:
+			// a low value in p2 bit will drive bit of p1 low when switch contact made
+		}
+		return 0xff;
+	});
```

Comment:
You should lift the `p2_r` out of the loop.

Reply:
ok, done